### PR TITLE
Simplify algorithm history comparison

### PR DESCRIPTION
**Pull request by quantumsteve**
Description of work.

This simplifies an overly expensive comparison function for `Mantid::API::AlgorithmHistory`. shared_ptrs are now passed by const reference instead of by value (I'd rather pass a simple reference, but the interface for `std::set` prevents that.) and removes the need to wrap the comparison function in a `boost::function`. I also updated a much simpler unnecessary `shared_ptr` copy in CompareWorkspaces.cpp. 

I found about a 10% performance improvement in WorkspaceHistoryTestPerformance ( before 0.53 sec, after 0.48 sec). 

The test itself spends most of its time parsing the defaultTime string.  Constructing a static object and calling the cheaper copy constructor brings the time to 0.17 sec.  Essentially, this performance test was primarily testing the DateAndTime constructor instead of WorkspaceHistory.

**To test:**

Code review: Can I (and if so should I) eliminate the template in CompareHistory? Any suggestions on what, if anything, to put in the release notes?

This is a small change with no issue number.

---
#### Reviewer

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):
##### Code Review
- [x] Is the code of an acceptable quality?
- [x] Does the code conform to the coding standards? Is it well structured with small focussed classes/methods/functions?
- [ ] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?
##### Functional Tests
- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?
- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/API/src/AlgorithmHistory.cpp [hunk 1]
```diff
@@ -28,22 +28,14 @@ AlgorithmHistory::AlgorithmHistory(const Algorithm *const alg,
                                    const double &duration,
                                    std::size_t uexeccount)
     : m_name(alg->name()), m_version(alg->version()), m_executionDate(start),
-      m_executionDuration(duration), m_execCount(uexeccount),
-      m_childHistories(boost::bind(CompareHistory::compare, _1, _2)) {
+      m_executionDuration(duration), m_execCount(uexeccount) {
```
Comment (martyngigg) [inline Framework/API/src/AlgorithmHistory.cpp]: Should we still initialize `m_childHistories` for completeness?

File: Framework/API/src/AlgorithmHistory.cpp [hunk 2]
```diff
@@ -60,8 +52,7 @@ AlgorithmHistory::AlgorithmHistory(const std::string &name, int vers,
                                    const double &duration,
                                    std::size_t uexeccount)
     : m_name(name), m_version(vers), m_executionDate(start),
-      m_executionDuration(duration), m_execCount(uexeccount),
-      m_childHistories(boost::bind(CompareHistory::compare, _1, _2)) {}
+      m_executionDuration(duration), m_execCount(uexeccount) {}
```
Comment (martyngigg) [inline Framework/API/src/AlgorithmHistory.cpp]: Same note as above.

**Comment by quantumsteve** [comment]
I added `m_childHistories()` to the member initializer lists. It shouldn't change the code, but I don't see any harm in being explicit.

**Comment by martyngigg** [comment]
I guess we could remove the template by having the comparator in a separate file, with the implementation in the `.cpp` and forward declares in the headers? While the template might be mildly confusing, the comment explains it well and moving introducing extra files seems unnecessary to me.

**Comment by martyngigg** [comment]
This is a good clean up that comes with a measurable performance gain. I'm happy to :shipit:

**Pull request merged**